crates/cxx-qt-gen/src/parser/constructor.rs

```diff
@@ -0,0 +1,126 @@
+// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+use syn::{
+    Path,
+    AngleBracketedGenericArguments, Error, ItemImpl,
+    PathArguments, PathSegment, Result,Type, GenericArgument,
+    spanned::Spanned, ImplItem
+};
+
+#[derive(Default)]
+pub struct ImplItems {
+    pub new_arguments: Option<Vec<Type>>,
+    pub base_arguments: Option<Vec<Type>>,
+    pub initialize_arguments: Option<Vec<Type>>,
+
+    pub passthrough: Vec<ImplItem>,
+}
+
+pub struct Constructor {
+    pub arguments: Vec<Type>,
+
+    pub items: ImplItems
+}
+
+impl Constructor {
+    fn parse_argument_list(ty:Type) -> Result<Vec<Type>>{
+        Ok(match ty {
+            Type::Tuple(tuple) => tuple.elems.into_iter().collect(),
+            _ => return Err(Error::new(ty.span(), "Expected a tuple as argument list!\nNote that a tuple of a single type needs to use a trailing comma, e.g. (i32,)"))
+        })
+    }
+
+    fn parse_impl_items(items: Vec<ImplItem>) -> Result<ImplItems> {
+        let mut result = ImplItems::default();
+
+        for item in items {
+            match item {
+                ImplItem::Type(impl_ty) => {
+                    match impl_ty.ident.to_string().as_str() {
+                        "NewArguments" => result.new_arguments = Some(Self::parse_argument_list(impl_ty.ty)?),
+                        "BaseArguments" => result.base_arguments = Some(Self::parse_argument_list(impl_ty.ty)?),
+                        "InitializeArguments" => result.initialize_arguments = Some(Self::parse_argument_list(impl_ty.ty)?),
+                        _ => result.passthrough.push(ImplItem::Type(impl_ty))
+                    }
+                }
+                _ => result.passthrough.push(item)
+            }
+        }
+
+        Ok(result)
+    }
+
+    fn parse_arguments(trait_path: &Path) -> Result<Vec<Type>> {
+        let constructor_path: Vec<_> = trait_path.segments.iter().collect();
+        if let [
+            // cxx_qt::
+            _cxx_qt, 
+            // Constructor
+            PathSegment {
+            arguments:
+                PathArguments::AngleBracketed(AngleBracketedGenericArguments { args: generics_punct, .. }),
+            ..
+        }] = *constructor_path
+        {
+            let generics : Vec<_> = generics_punct.iter().collect();
+            if generics.len() != 1 {
+                // if the generics are empty, using the span of `generics_punct` causes an invalid
+                // span to be returned!
+                let span = if generics.is_empty() { trait_path.span() } else { generics_punct.span() };
+                return Err(Error::new(span, "cxx_qt::Constructor expects a single generic argument, use a tuple to provide multiple arguments!"))
+            }
+
+            if let [GenericArgument::Type(ty)] = *generics {
```

Comment:
the `[...]` is curious here, is that pattern matching on the first element in a slice / array ?

Reply:
yes it is and it's super useful.
Got inspired by this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment?retiredLocale=de
Turns out the same works in Rust.